pkg/cache/resource_test.go

```diff
@@ -0,0 +1,292 @@
+				// its available, to prevent
+				// overadmission for (red,cpu) within
+				// the CohortTree.
+				"cq2": {{Flavor: "red", Resource: "cpu"}: 0},
+			},
+			wantPotentiallyAvailable: map[string]resources.FlavorResourceQuantities{
+				"cq1": {{Flavor: "red", Resource: "cpu"}: 20_000},
+				"cq2": {{Flavor: "red", Resource: "cpu"}: 10_000},
+			},
+		},
+		"lending and borrowing limits respected": {
+			// cohort has 40 (red, cpu) total
+			// cq1 can access 35 = 40 - 5 reserved
+			// cq2 can access 12, due to borrowing limit
+			// cq3 can access 40, as there are no restrictions.
+			clusterQueues: []kueue.ClusterQueue{
+				utiltesting.MakeClusterQueue("cq1").
+					Cohort("cohort").
+					ResourceGroup(
+						*utiltesting.MakeFlavorQuotas("red").Resource("cpu", "10").Obj(),
+					).ClusterQueue,
+				utiltesting.MakeClusterQueue("cq2").
+					Cohort("cohort").
+					ResourceGroup(
+						*utiltesting.MakeFlavorQuotas("red").Resource("cpu", "10", "2").Obj(),
+					).ClusterQueue,
+				utiltesting.MakeClusterQueue("cq3").
+					Cohort("cohort").
+					ResourceGroup(
+						*utiltesting.MakeFlavorQuotas("red").Resource("cpu", "10", "", "5").Obj(),
+					).ClusterQueue,
+			},
+			cohorts: []kueuealpha.Cohort{
+				utiltesting.MakeCohort("cohort").
+					ResourceGroup(
+						*utiltesting.MakeFlavorQuotas("red").Resource("cpu", "10").Obj(),
+					).Cohort,
+			},
+			usage: map[string]resources.FlavorResourceQuantities{
+				"cq1": {{Flavor: "red", Resource: "cpu"}: 20_000},
+				"cq2": {{Flavor: "red", Resource: "cpu"}: 10_000},
+				"cq3": {{Flavor: "red", Resource: "cpu"}: 6_000},
+			},
+			wantAvailable: map[string]resources.FlavorResourceQuantities{
+				// cq1 uses 20k, cq2 uses 10k, and
+				// cq3 uses 1k (not counting first 5k
+				// in guaranteedQuota), resulting in
+				// 4k left of potential 35k.
+				"cq1": {{Flavor: "red", Resource: "cpu"}: 4_000},
+				// cq2 has access to only 2k of the
+				// remaining 4k due to its borrowing
+				// limit.
+				"cq2": {{Flavor: "red", Resource: "cpu"}: 2_000},
+				// 40k - (20k + 10k + 6k) = 4k
+				"cq3": {{Flavor: "red", Resource: "cpu"}: 4_000},
+			},
+			wantPotentiallyAvailable: map[string]resources.FlavorResourceQuantities{
+				"cq1": {{Flavor: "red", Resource: "cpu"}: 35_000},
+				"cq2": {{Flavor: "red", Resource: "cpu"}: 12_000},
+				"cq3": {{Flavor: "red", Resource: "cpu"}: 40_000},
+			},
+		},
+	}
+
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			ctx := context.Background()
+			cache := New(utiltesting.NewFakeClient())
+			cache.AddOrUpdateResourceFlavor(utiltesting.MakeResourceFlavor("red").Obj())
+			cache.AddOrUpdateResourceFlavor(utiltesting.MakeResourceFlavor("blue").Obj())
+			for _, cq := range tc.clusterQueues {
+				_ = cache.AddClusterQueue(ctx, &cq)
+			}
+			for _, cohort := range tc.cohorts {
+				cache.AddCohort(&cohort)
+			}
+
+			snapshot := cache.Snapshot()
+
+			t.Run("before adding usage", func(t *testing.T) {
```

Comment:
remove `t.Run` here, as the remaining subtests depend on it.

Reply:
Removed all 3, but left the blocks with a comment, so that the `got` variables are scoped.